frontend/src/widgets/CoursesAssociatedWithActivityReports.js

```diff
@@ -0,0 +1,230 @@
+import { COURSES_PER_PAGE } from '../Constants';
+
+export const parseValue = (value) => {
+  const noCommasValue = value.replaceAll(',', '');
+  const parsedValue = parseInt(noCommasValue, DECIMAL_BASE);
+  if (Number.isNaN(parsedValue)) {
+    return value;
+  }
+  return parsedValue;
+};
+function CoursesAssociatedWithActivityReports({
+  data,
+  loading,
+  resetPagination,
+  setResetPagination,
+  perPageNumber,
+}) {
+  const [courseUse, setCourseUse] = useState([]);
+  const [courseCount, setCourseCount] = useState(0);
+  const [localLoading, setLocalLoading] = useState(false);
+  const [perPage] = useState(perPageNumber);
+  const [sortConfig, setSortConfig] = useSessionSort({
+    sortBy: '1',
+    direction: 'desc',
+    activePage: 1,
+  }, 'activityReportsTable');
+  const [coursesPerPage, setCoursesPerPage] = useState([]);
+  const [checkBoxes, setCheckBoxes] = useState({});
+
+  const { activePage } = sortConfig;
+
+  const [offset, setOffset] = useState((activePage - 1) * perPage);
+
+  useEffect(() => {
+    try {
+      // Set local data.
+      setLocalLoading(true);
+      const courseToUse = data.courses || [];
+      setCourseUse(courseToUse);
+      setCourseCount(courseToUse.length);
+    } finally {
+      setLocalLoading(false);
+    }
+  }, [data]);
+
+  // a side effect that resets the pagination when the filters change
+  useEffect(() => {
+    if (resetPagination) {
+      setSortConfig({ ...sortConfig, activePage: 1 });
+      setOffset(0); // 0 times perpage = 0
+      setResetPagination(false);
+    }
+  }, [activePage, perPage, resetPagination, setResetPagination, setSortConfig, sortConfig, data]);
+
+  const handlePageChange = (pageNumber) => {
+    if (!loading) {
+      // copy state
+      const sort = { ...sortConfig };
+
+      // mutate
+      sort.activePage = pageNumber;
+
+      // store it
+      setSortConfig(sort);
+      setOffset((pageNumber - 1) * perPage);
+    }
+  };
+
+  const requestSort = (sortBy) => {
+    // Get sort direction.
+    let direction = 'asc';
+    if (
+      sortConfig
+      && sortConfig.sortBy === sortBy
+      && sortConfig.direction === 'asc'
+    ) {
+      direction = 'desc';
+    }
+
+    // Set the value we want to sort by.
+    const valuesToSort = sortBy === 'Course'
+      ? courseUse.map((t) => ({
+        ...t,
+        sortBy: t.heading,
+      }))
+      : courseUse.map((t) => (
+        {
+          ...t,
+          sortBy: parseValue(t.data.find((tp) => tp.title === sortBy).value),
+        }));
+
+    // Value sort.
+    const sortValueA = direction === 'asc' ? 1 : -1;
+    const sortValueB = direction === 'asc' ? -1 : 1;
+    valuesToSort.sort((a, b) => {
+      if (a.sortBy > b.sortBy) {
+        return sortValueA;
+      } if (b.sortBy > a.sortBy) {
+        return sortValueB;
+      }
+      return 0;
+    });
+
+    setCourseUse(valuesToSort);
+    setOffset(0);
+    setSortConfig({ sortBy, direction, activePage: 1 });
+  };
+
```

Comment:
Maybe something for a different PR, but we write this code over and over again. An excellent candidate for abstracting into a hook. Same with that CSV code I think.

Reply:
Probably warrants its own ticket but a good idea.